src/blockMonitor/blockTimeoutDetector.ts

```diff
@@ -0,0 +1,58 @@
+import { BlockProcessor } from "./blockProcessor";
+import { StartStopService } from "../dataEntities";
+
+/**
+ * Generates events when no new block is observed for too long, possibly signaling a malfunctioning of the provider.
+ * The timer starts when the service is started, and is reset every time a new block is received.
+ */
+export class BlockTimeoutDetector extends StartStopService {
+    private timeoutHandler: NodeJS.Timeout | null = null;
+
+    /**
+     * Emitted when a no new block is received for a time longer than the `timeout` milliseconds.
+     */
+    public static readonly BLOCK_TIMEOUT_EVENT = "no_new_block";
+
+    /**
+     * @param timeout The number of milliseconds without a new block before generating
+     */
+    constructor(private blockProcessor: BlockProcessor, public readonly timeout: number) {
+        super("Block timeout detector");
+        this.handleNewBlock = this.handleNewBlock.bind(this);
+        this.handleNoNewBlockTimeout = this.handleNoNewBlockTimeout.bind(this);
+    }
+
+    protected async startInternal(): Promise<void> {
+        this.blockProcessor.on(BlockProcessor.NEW_HEAD_EVENT, this.handleNewBlock);
+        this.initNoNewBlockTimeout();
+    }
+
+    public async stopInternal(): Promise<void> {
+        this.clearNoNewBlockTimeout();
+        this.blockProcessor.off(BlockProcessor.NEW_HEAD_EVENT, this.handleNewBlock);
+    }
+
+    // Start the timer, after clearing any previous timer
+    private initNoNewBlockTimeout() {
```

Comment:
I think setNoNewBlockTimeout is a slightly clearer name

Reply:
This should be wrapped in a try/catch

Reply:
Why? It cannot throw an exception, as far as I can tell. It is only clearing a timeout handler (if any), and calling a new setTimeout

Renamed to **reset**NoNewBlockTimeout, since it possibly also clears the existing timeout. (f862b17adf715ae2db894a21a4292a1ac9912b62)

Reply:
Yeah, I think we had a chat about this once before. I feel like it's error prone either way.

If you don't add a try/catch someone may to edit your logic, make it more complex and not think to add the try/catch which they now need to do.

If you add a try/catch it's unnecessary so someone it's likely to just remove it, even if you comment it. In which case you're back at square one. In the mean time, you've also added unnecessary code.

What I was kinda thinking was that we could just make a manual global rule which is: if it's an event handler, wrap it in try/catch. But it's a human process, so it probably wont work either. What dyou think? I also don't mind if you don't add the try/catch, since none of the above solutions are good ones.

Reply:
Yeah, generally speaking, I would say that _all_ code should catch potential exceptions, if they can happen; it would be good if that could be enforced at compile-level, but we're not that lucky with TypeScript.
One option could be to use a decorator for all event handlers to handle the logic of wrapping in a try/catch and logging any error, but they are still an experimental feature, not sure we want to use them: https://www.typescriptlang.org/docs/handbook/decorators.html

Reply:
I like the idea of a decorator as it would be declare it's use case: @EventHandler

We can do this/investigate in a different ticket for now. Do what you want for the case at hand.